File: ruby/fsrouter.rb
#!/usr/bin/env ruby
require 'json'
require 'find'
require 'open3'
require 'timeout'
require 'uri'
require 'webrick'

HTTP_METHODS = %w[GET HEAD POST PUT DELETE PATCH OPTIONS].freeze

class Node
  attr_accessor :literal, :param, :param_name, :handlers

  def initialize(param_name = '')
    @literal = {}
    @param = nil
    @param_name = param_name
    @handlers = {}
  end

  def match(segs)
    params = {}
    cur = self
    segs.each do |seg|
      if cur.literal.key?(seg)
        cur = cur.literal[seg]
      elsif cur.param
        params[cur.param.param_name] = seg
        cur = cur.param
      else
        return [nil, nil]
      end
    end
    [cur, params]
  end
end

def normalize_request_path(path)
  collapsed = path.gsub(%r{/+}, '/')
  collapsed = '/' if collapsed.empty?
  trimmed = collapsed == '/' ? collapsed : collapsed.sub(%r{/$}, '')
  segs = []
  trimmed.split('/').each do |segment|
    next if segment.empty?
    decoded = URI::DEFAULT_PARSER.unescape(segment)
    raise ArgumentError, 'invalid path' if decoded == '..'
    segs << decoded
  end
  segs
end

def is_executable(path)
  File.file?(path) && File.executable?(path)
end

def join_prefix(prefix, seg)
  prefix.empty? ? seg : "#{prefix}/#{seg}"
end

def build_tree(route_dir)
  abs_dir = File.realpath(route_dir)
  root = Node.new

  Find.find(abs_dir) do |path|
    next if File.directory?(path)

    method = File.basename(path).upcase
    next unless HTTP_METHODS.include?(method)

    rel_dir = File.dirname(path).delete_prefix(abs_dir)
    segs = rel_dir.split('/').reject(&:empty?)
    cur = root
    segs.each do |seg|
      if seg.start_with?(':')
        cur.param ||= Node.new(seg[1..-1])
        cur = cur.param
      else
        cur.literal[seg] ||= Node.new
        cur = cur.literal[seg]
      end
    end
    cur.handlers[method] = path
  end

  root
end

def collect_routes(node, prefix, items)
  route = prefix.empty? ? '/' : "/#{prefix}"
  node.handlers.each do |method, path|
    tag = begin
      is_executable(path) ? 'exec' : 'static'
    rescue StandardError
      'unknown'
    end
    items << [route, method, path, tag]
  end
  node.literal.keys.sort.each do |seg|
    collect_routes(node.literal[seg], join_prefix(prefix, seg), items)
  end
  collect_routes(node.param, join_prefix(prefix, ":#{node.param.param_name}"), items) if node.param
end

def print_routes(root, route_dir)
  $stderr.puts("routes from #{route_dir}:")
  items = []
  collect_routes(root, '', items)
  items.sort_by { |item| [item[0], item[1]] }.each do |route, method, path, tag|
    $stderr.puts(format('  %-7s %-45s → %s [%s]', method, route, path, tag))
  end
  $stderr.flush
end

def env_key(value)
  value.upcase.tr('-', '_')
end

def split_host_port(value)
  return ['', ''] if value.nil? || value.empty?

  if value.start_with?('[') && value.include?(']')
    close = value.index(']')
    host = value[1...close]
    rest = value[(close + 1)..]
    return [host, rest.start_with?(':') ? rest[1..-1] : '']
  end

  if value.count(':') == 1
    host, port = value.split(':', 2)
    return [host, port]
  end

  [value, '']
end

def build_env(req, server, params)
  env = ENV.to_h
  env['REQUEST_METHOD'] = req.request_method
  env['REQUEST_URI'] = req.request_uri.to_s
  env['REQUEST_PATH'] = req.path
  env['QUERY_STRING'] = req.query_string.to_s
  env['CONTENT_TYPE'] = req['content-type'].to_s
  env['CONTENT_LENGTH'] = req['content-length'].to_s
  env['REMOTE_ADDR'] = "#{req.peeraddr[3]}:#{req.peeraddr[1]}"

  server_name, server_port = split_host_port(req.host || server[:listen_addr])
  env['SERVER_NAME'] = server_name
  env['SERVER_PORT'] = server_port unless server_port.empty?

  params.each do |key, value|
    env["PARAM_#{env_key(key)}"] = value
  end

  seen_query = {}
  URI.decode_www_form(req.query_string.to_s).each do |key, value|
    next if seen_query[key]
    seen_query[key] = true
    env["QUERY_#{env_key(key)}"] = value
  end

  seen_headers = {}
  req.each do |key, value|
    next if seen_headers[key]
    seen_headers[key] = true
    env["HTTP_#{key.upcase.tr('-', '_')}"] = value.to_s
  end

  env
end

def looks_like_header(raw)
  raw.each_byte.with_index do |byte, idx|
    return true if byte == ':'.ord && idx.positive?
    return false if ["\n".ord, "\r".ord, ' '.ord, "\t".ord].include?(byte)
  end
  false
end

def parse_header_line(line)
  idx = line.index(':')
  return nil unless idx && idx.positive?

  key = line[0...idx]
  return nil if key.each_char.any? { |ch| ch.ord <= 32 || ch.ord == 127 }

  [key, line[(idx + 1)..].strip]
end

def parse_cgi_headers(raw, default_status)
  status = default_status
  content_type = 'application/json'
  headers = []
  body_index = nil
  lines = raw.split(/\n/, -1)
  offset = 0

  lines.each do |line|
    consumed = line.bytesize + 1
    line = line.delete_suffix("\r")
    if line.empty?
      body_index = offset + consumed
      break
    end

    parsed = parse_header_line(line.force_encoding('UTF-8'))
    return nil unless parsed

    key, value = parsed
    if key.downcase == 'status'
      first = value.split.first
      status = first.to_i if first && first.match?(/^\d+$/)
    elsif key.downcase == 'content-type'
      content_type = value
    else
      headers << [key, value]
    end
    offset += consumed
  end

  return nil unless body_index

  [status, content_type, headers, raw.byteslice(body_index..) || '']
end

def exit_to_status(code)
  return 200 if code == 0
  return 400 if code == 1

  500
end

def json_response(res, req, status, payload)
  body = JSON.generate(payload)
  res.status = status
  res['Content-Type'] = 'application/json'
  res['Content-Length'] = body.bytesize.to_s
  res.body = req.request_method == 'HEAD' ? '' : body
end

def serve_static(req, res, handler_path)
  data = File.binread(handler_path)
  res.status = 200
  res['Content-Type'] = WEBrick::HTTPUtils.mime_type(handler_path, WEBrick::HTTPUtils::DefaultMimeTypes)
  res['Content-Length'] = data.bytesize.to_s
  res.body = req.request_method == 'HEAD' ? '' : data
  200
rescue StandardError => e
  json_response(res, req, 500, { error: 'static_read_failed', message: e.message })
  500
end

def execute_handler(req, res, server, handler_path, params)
  env = build_env(req, server, params)
  body = req.body || ''
  stdout = ''
  stderr = ''
  status = nil
  wait_thr = nil
  pid = nil

  begin
    Open3.popen3(env, handler_path, chdir: File.dirname(handler_path)) do |stdin, out, err, thr|
      wait_thr = thr
      pid = thr.pid
      stdin.binmode
      out.binmode
      err.binmode
      stdin.write(body)
      stdin.close
      begin
        Timeout.timeout(server[:command_timeout]) do
          stdout_reader = Thread.new { out.read }
          stderr_reader = Thread.new { err.read }
          status = thr.value
          stdout = stdout_reader.value
          stderr = stderr_reader.value
        end
      rescue Timeout::Error
        begin
          Process.kill('TERM', pid)
        rescue StandardError
        end
        begin
          Timeout.timeout(1) { thr.value }
        rescue StandardError
          begin
            Process.kill('KILL', pid)
          rescue StandardError
          end
        end
        json_response(res, req, 504, { error: 'handler_timeout', timeout_seconds: server[:command_timeout] })
        return 504
      end
    end
  rescue Errno::ENOENT, Errno::EACCES => e
    json_response(res, req, 502, { error: 'exec_failed', message: e.message })
    return 502
  end

  unless stderr.empty?
    $stderr.puts("  [handler stderr] #{stderr.encode('UTF-8', invalid: :replace, undef: :replace).rstrip}")
    $stderr.flush
  end

  exit_code = status&.exitstatus || 0
  raw = stdout
  raw = stderr if raw.empty? && exit_code != 0 && !stderr.empty?
  write_cgi_response(req, res, raw, exit_code)
end

def write_cgi_response(req, res, raw, exit_code)
  status = exit_to_status(exit_code)
  content_type = 'application/json'
  headers = []
  body = raw

  if !raw.empty? && looks_like_header(raw)
    parsed = parse_cgi_headers(raw, status)
    if parsed
      status, content_type, headers, body = parsed
    end
  end

  res.status = status
  headers.each { |key, value| res[key] = value }
  res['Content-Type'] = content_type
  res['Content-Length'] = body.bytesize.to_s
  res.body = req.request_method == 'HEAD' ? '' : body
  status
end

def handle_handler(req, res, server, handler_path, params)
  return serve_static(req, res, handler_path) unless is_executable(handler_path)

  execute_handler(req, res, server, handler_path, params)
rescue StandardError => e
  json_response(res, req, 500, { error: 'handler_stat_failed', message: e.message })
  500
end

def parse_listen_addr(addr)
  return ['0.0.0.0', addr[1..].to_i] if addr.start_with?(':')
  if addr.start_with?('[') && addr.include?(']')
    host, port = split_host_port(addr)
    return [host, port.to_i]
  end
  if addr.count(':') == 1
    host, port = addr.split(':', 2)
    return [host, port.to_i]
  end
  [addr, 8080]
end

def env_or(key, fallback)
  value = ENV[key]
  value.nil? || value.empty? ? fallback : value
end

def process_request(req, res, root, server_config)
  start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  status = 500

  begin
    segs = normalize_request_path(req.path)
    node, params = root.match(segs)
    if node.nil? || node.handlers.empty?
      json_response(res, req, 404, { error: 'not_found', path: req.path })
      status = 404
    else
      handler_path = node.handlers[req.request_method]
      handler_path ||= node.handlers['GET'] if req.request_method == 'HEAD'
      if handler_path.nil?
        allowed = node.handlers.keys.sort
        json_response(res, req, 405, { error: 'method_not_allowed', allow: allowed })
        res['Allow'] = allowed.join(', ')
        status = 405
      else
        status = handle_handler(req, res, server_config, handler_path, params || {})
      end
    end
  rescue ArgumentError
    json_response(res, req, 400, { error: 'invalid_path', path: req.path })
    status = 400
  end

  elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
  $stderr.puts(format('%s %s → %d (%.6fs)', req.request_method, req.path, status, elapsed))
  $stderr.flush
end

class FsrouterServlet < WEBrick::HTTPServlet::AbstractServlet
  def initialize(server, root, server_config)
    super(server)
    @root = root
    @server_config = server_config
  end

  def do_GET(req, res)
    process_request(req, res)
  end

  def do_HEAD(req, res)
    process_request(req, res)
  end

  def do_POST(req, res)
    process_request(req, res)
  end

  def do_PUT(req, res)
    process_request(req, res)
  end

  def do_DELETE(req, res)
    process_request(req, res)
  end

  def do_PATCH(req, res)
    process_request(req, res)
  end

  def do_OPTIONS(req, res)
    process_request(req, res)
  end

  private

  def process_request(req, res)
    Object.new.send(:process_request, req, res, @root, @server_config)
  end
end

route_dir = env_or('ROUTE_DIR', './routes')
listen_addr = env_or('LISTEN_ADDR', ':8080')
timeout_seconds = Integer(env_or('COMMAND_TIMEOUT', '30'), exception: false) || 30
timeout_seconds = 30 if timeout_seconds <= 0

begin
  root = build_tree(route_dir)
rescue StandardError => e
  $stderr.puts("failed to scan #{route_dir}: #{e.message}")
  $stderr.flush
  exit 1
end

print_routes(root, route_dir)
host, port = parse_listen_addr(listen_addr)
server_config = { root: root, command_timeout: timeout_seconds, listen_addr: listen_addr }

server = WEBrick::HTTPServer.new(
  BindAddress: host,
  Port: port,
  AccessLog: [],
  Logger: WEBrick::Log.new($stderr, WEBrick::Log::WARN)
)

trap('INT') { server.shutdown }
trap('TERM') { server.shutdown }

server.mount('/', FsrouterServlet, root, server_config)

$stderr.puts("listening on #{listen_addr} (timeout #{timeout_seconds}s)")
$stderr.flush
server.start

File: examples/claude-as-anthropic-api-server/fsrouter.py
#!/usr/bin/env python3
import json
import mimetypes
import os
import signal
import stat
import subprocess
import sys
import threading
import time
import urllib.parse
from dataclasses import dataclass, field
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from typing import Optional

HTTP_METHODS = {"GET", "HEAD", "POST", "PUT", "DELETE", "PATCH", "OPTIONS"}


@dataclass
class Node:
    literal: dict[str, "Node"] = field(default_factory=dict)
    param: Optional["Node"] = None
    param_name: str = ""
    handlers: dict[str, Path] = field(default_factory=dict)
    implicit_handler: Optional[Path] = None

    def match(self, segs: list[str]) -> tuple[Optional["Node"], Optional[dict[str, str]]]:
        params: dict[str, str] = {}
        cur = self
        for seg in segs:
            if seg in cur.literal:
                cur = cur.literal[seg]
            elif cur.param is not None:
                params[cur.param.param_name] = seg
                cur = cur.param
            else:
                return None, None
        return cur, params


class FsrouterServer(ThreadingHTTPServer):
    daemon_threads = True

    def __init__(self, server_address: tuple[str, int], handler_class, root: Node, timeout_seconds: int, listen_addr: str, route_dir_abs: Path):
        super().__init__(server_address, handler_class)
        self.root = root
        self.command_timeout = timeout_seconds
        self.listen_addr = listen_addr
        self.route_dir_abs = route_dir_abs


class Handler(BaseHTTPRequestHandler):
    server: FsrouterServer

    def do_GET(self):
        self.handle_method()

    def do_HEAD(self):
        self.handle_method()

    def do_POST(self):
        self.handle_method()

    def do_PUT(self):
        self.handle_method()

    def do_DELETE(self):
        self.handle_method()

    def do_PATCH(self):
        self.handle_method()

    def do_OPTIONS(self):
        self.handle_method()

    def handle_method(self):
        start = time.time()
        parsed = urllib.parse.urlsplit(self.path)

        try:
            segs = normalize_request_path(parsed.path)
        except ValueError:
            self.write_json(400, {"error": "invalid_path", "path": parsed.path})
            self.log_result(400, start)
            return

        node, params = self.server.root.match(segs)
        if node is None or (not node.handlers and node.implicit_handler is None):
            status = self.serve_filesystem_fallback(segs, parsed.path)
            self.log_result(status, start)
            return

        handler_path = node.handlers.get(self.command)
        if handler_path is not None:
            status = self.handle_handler(handler_path, params or {})
            self.log_result(status, start)
            return

        if node.implicit_handler is not None:
            status = self.handle_handler(node.implicit_handler, params or {})
            self.log_result(status, start)
            return

        allowed = sorted(node.handlers)
        body = json.dumps({"error": "method_not_allowed", "allow": allowed}, separators=(",", ":")).encode()
        self.send_response(405)
        self.send_header("Content-Type", "application/json")
        self.send_header("Allow", ", ".join(allowed))
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        if self.command != "HEAD":
            self.wfile.write(body)
        self.log_result(405, start)

    def handle_handler(self, handler_path: Path, params: dict[str, str]) -> int:
        try:
            st = handler_path.stat()
        except OSError as err:
            self.write_json(500, {"error": "handler_stat_failed", "message": str(err)})
            return 500

        if not is_executable(st):
            return self.serve_static(handler_path)

        return self.execute_handler(handler_path, params)

    def serve_filesystem_fallback(self, segs: list[str], request_path: str) -> int:
        fallback = self.server.route_dir_abs.joinpath(*segs) if segs else self.server.route_dir_abs
        if fallback.is_file():
            return self.serve_fallback_file(fallback)
        if fallback.is_dir():
            preferred = self.find_directory_index(fallback)
            if preferred is not None:
                return self.serve_fallback_file(preferred)
            return self.serve_dir_listing(fallback, request_path)
        self.write_json(404, {"error": "not_found", "path": request_path})
        return 404

    def find_directory_index(self, dir_path: Path) -> Optional[Path]:
        for name in ("index.html", "index.htm"):
            path = dir_path / name
            if path.is_file():
                return path
        executable_indexes: list[Path] = []
        try:
            for entry in dir_path.iterdir():
                if not entry.is_file() or not entry.name.startswith("index."):
                    continue
                if is_executable(entry.stat()):
                    executable_indexes.append(entry)
        except OSError:
            return None
        if executable_indexes:
            return sorted(executable_indexes, key=lambda p: p.name)[0]
        return None

    def serve_fallback_file(self, path: Path) -> int:
        try:
            st = path.stat()
        except OSError as err:
            self.write_json(500, {"error": "handler_stat_failed", "message": str(err)})
            return 500
        if is_executable(st):
            return self.execute_handler(path, {})
        return self.serve_static(path)

    def execute_handler(self, path: Path, params: dict[str, str]) -> int:
        length = int(self.headers.get("Content-Length", "0") or "0")
        request_body = self.rfile.read(length) if length > 0 else b""
        env = build_env(self, params)

        try:
            proc = subprocess.Popen(
                [str(path)],
                cwd=str(path.parent),
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                env=env,
            )
        except OSError as err:
            self.write_json(502, {"error": "exec_failed", "message": str(err)})
            return 502

        try:
            stdout, stderr = proc.communicate(request_body, timeout=self.server.command_timeout)
        except subprocess.TimeoutExpired:
            proc.kill()
            proc.communicate()
            self.write_json(504, {"error": "handler_timeout", "timeout_seconds": self.server.command_timeout})
            return 504

        if stderr:
            sys.stderr.write(f"  [handler stderr] {stderr.decode('utf-8', errors='replace').rstrip()}\n")
            sys.stderr.flush()

        exit_code = proc.returncode or 0
        body = stdout
        if not body and exit_code != 0 and stderr:
            body = stderr

        status = exit_to_status(exit_code)
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        if self.command != "HEAD":
            self.wfile.write(body)
        return status

    def serve_dir_listing(self, dir_path: Path, request_path: str) -> int:
        try:
            entries = sorted(dir_path.iterdir(), key=lambda e: (not e.is_dir(), e.name.lower()))
        except OSError as err:
            self.write_json(500, {"error": "dir_listing_failed", "message": str(err)})
            return 500
        title = f"Index of {request_path or '/'}"
        lines = [f"<!DOCTYPE html><html><head><title>{title}</title></head><body><h1>{title}</h1><ul>"]
        if request_path and request_path != "/":
            lines.append("<li><a href=\"../\">../</a></li>")
        for entry in entries:
            name = entry.name
            if entry.is_dir():
                lines.append(f"<li><a href=\"{name}/\">{name}/</a></li>")
            else:
                lines.append(f"<li><a href=\"{name}\">{name}</a></li>")
        lines.append("</ul></body></html>")
        body = "\n".join(lines).encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        if self.command != "HEAD":
            self.wfile.write(body)
        return 200

    def serve_static(self, handler_path: Path) -> int:
        try:
            data = handler_path.read_bytes()
        except OSError as err:
            self.write_json(500, {"error": "static_read_failed", "message": str(err)})
            return 500

        content_type = mimetypes.guess_type(str(handler_path))[0] or "application/octet-stream"
        self.send_response(200)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        if self.command != "HEAD":
            self.wfile.write(data)
        return 200

    def write_json(self, status: int, payload: dict) -> None:
        body = json.dumps(payload, separators=(",", ":")).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        if self.command != "HEAD":
            self.wfile.write(body)

    def log_message(self, fmt: str, *args):
        sys.stderr.write("%s - - [%s] %s\n" % (self.address_string(), self.log_date_time_string(), fmt % args))

    def log_result(self, status: int, start: float):
        elapsed = time.time() - start
        sys.stderr.write(f"{self.command} {urllib.parse.urlsplit(self.path).path} → {status} ({elapsed:.6f}s)\n")
        sys.stderr.flush()


def normalize_request_path(path: str) -> list[str]:
    segs: list[str] = []
    for segment in path.split("/"):
        if not segment:
            continue
        decoded = urllib.parse.unquote(segment)
        if decoded == "..":
            raise ValueError("invalid path")
        segs.append(decoded)
    return segs


def build_tree(route_dir: str) -> Node:
    root = Node()
    abs_dir = Path(route_dir).resolve(strict=True)
    for current_root, _, files in os.walk(abs_dir, followlinks=True):
        current_root_path = Path(current_root)
        for filename in files:
            file_path = current_root_path / filename
            method = filename.upper()
            if method in HTTP_METHODS:
                rel = file_path.parent.relative_to(abs_dir)
                segs = [segment for segment in rel.as_posix().split("/") if segment and segment != "."]
                cur = root
                for seg in segs:
                    if seg.startswith(":"):
                        if cur.param is None:
                            cur.param = Node(param_name=seg[1:])
                        cur = cur.param
                    else:
                        cur = cur.literal.setdefault(seg, Node())
                cur.handlers[method] = file_path
            else:
                try:
                    if not is_executable(file_path.stat()):
                        continue
                except OSError:
                    continue
                rel = file_path.relative_to(abs_dir)
                segs = [segment for segment in rel.as_posix().split("/") if segment and segment != "."]
                cur = root
                for seg in segs:
                    if seg.startswith(":"):
                        if cur.param is None:
                            cur.param = Node(param_name=seg[1:])
                        cur = cur.param
                    else:
                        cur = cur.literal.setdefault(seg, Node())
                cur.implicit_handler = file_path
    return root


def collect_routes(node: Node, prefix: str, items: list[tuple[str, str, Path, str]]) -> None:
    route = "/" if not prefix else f"/{prefix}"
    for method, path in node.handlers.items():
        try:
            tag = "exec" if is_executable(path.stat()) else "static"
        except OSError:
            tag = "unknown"
        items.append((route, method, path, tag))
    if node.implicit_handler is not None:
        items.append((route, "*", node.implicit_handler, "exec"))
    for seg in sorted(node.literal):
        collect_routes(node.literal[seg], join_prefix(prefix, seg), items)
    if node.param is not None:
        collect_routes(node.param, join_prefix(prefix, f":{node.param.param_name}"), items)


def join_prefix(prefix: str, seg: str) -> str:
    return seg if not prefix else f"{prefix}/{seg}"


def print_routes(root: Node, route_dir: str) -> None:
    sys.stderr.write(f"routes from {route_dir}:\n")
    items: list[tuple[str, str, Path, str]] = []
    collect_routes(root, "", items)
    items.sort(key=lambda item: (item[0], item[1]))
    for route, method, path, tag in items:
        sys.stderr.write(f"  {method:<7} {route:<45} → {path} [{tag}]\n")
    sys.stderr.flush()


def build_env(handler: Handler, params: dict[str, str]) -> dict[str, str]:
    parsed = urllib.parse.urlsplit(handler.path)
    env = dict(os.environ)
    env["REQUEST_METHOD"] = handler.command
    env["REQUEST_URI"] = handler.path
    env["REQUEST_PATH"] = parsed.path
    env["QUERY_STRING"] = parsed.query
    env["CONTENT_TYPE"] = handler.headers.get("Content-Type", "")
    env["CONTENT_LENGTH"] = handler.headers.get("Content-Length", "")
    env["REMOTE_ADDR"] = f"{handler.client_address[0]}:{handler.client_address[1]}"

    server_name, server_port = split_host_port(handler.headers.get("Host") or handler.server.listen_addr)
    env["SERVER_NAME"] = server_name
    if server_port:
        env["SERVER_PORT"] = server_port

    for key, value in params.items():
        env[f"PARAM_{env_key(key)}"] = value

    seen_query: set[str] = set()
    for key, value in urllib.parse.parse_qsl(parsed.query, keep_blank_values=True):
        if key not in seen_query:
            seen_query.add(key)
            env[f"QUERY_{env_key(key)}"] = value

    seen_headers: set[str] = set()
    for key in handler.headers:
        lower = key.lower()
        if lower in seen_headers:
            continue
        seen_headers.add(lower)
        env[f"HTTP_{key.upper().replace('-', '_')}"] = handler.headers.get(key, "")

    return env


def split_host_port(value: str) -> tuple[str, str]:
    if not value:
        return "", ""
    if value.startswith("[") and "]" in value:
        end = value.find("]")
        host = value[1:end]
        rest = value[end + 1 :]
        if rest.startswith(":"):
            return host, rest[1:]
        return host, ""
    if value.count(":") == 1:
        host, port = value.rsplit(":", 1)
        return host, port
    return value, ""


def env_key(value: str) -> str:
    return value.upper().replace("-", "_")


def exit_to_status(code: int) -> int:
    if code == 0:
        return 200
    if code == 1:
        return 400
    return 500


def is_executable(st: os.stat_result) -> bool:
    return bool(st.st_mode & (stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH))


def env_or(key: str, fallback: str) -> str:
    value = os.environ.get(key)
    return value if value else fallback


def parse_listen_addr(addr: str) -> tuple[str, int]:
    if addr.startswith(":"):
        return "", int(addr[1:])
    if addr.startswith("[") and "]" in addr:
        host, port = split_host_port(addr)
        return host, int(port)
    if addr.count(":") == 1:
        host, port = addr.rsplit(":", 1)
        return host, int(port)
    return addr, 8080


def main() -> int:
    route_dir = env_or("ROUTE_DIR", "./routes")
    listen_addr = env_or("LISTEN_ADDR", ":8080")
    try:
        timeout_seconds = int(env_or("COMMAND_TIMEOUT", "30"))
    except ValueError:
        timeout_seconds = 30
    if timeout_seconds <= 0:
        timeout_seconds = 30

    try:
        root = build_tree(route_dir)
    except OSError as err:
        sys.stderr.write(f"failed to scan {route_dir}: {err}\n")
        sys.stderr.flush()
        return 1

    print_routes(root, route_dir)
    host, port = parse_listen_addr(listen_addr)
    server = FsrouterServer((host, port), Handler, root, timeout_seconds, listen_addr, Path(route_dir).resolve())

    def shutdown_handler(signum, frame):
        sys.stderr.write("shutting down...\n")
        sys.stderr.flush()
        threading.Thread(target=server.shutdown, daemon=True).start()

    for sig in (signal.SIGINT, signal.SIGTERM):
        signal.signal(sig, shutdown_handler)

    bound_host, bound_port = server.server_address[:2]
    display_addr = listen_addr if listen_addr else f"{bound_host}:{bound_port}"
    sys.stderr.write(f"listening on {display_addr} (timeout {timeout_seconds}s)\n")
    sys.stderr.flush()

    try:
        server.serve_forever()
    finally:
        server.server_close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
